<h2> Calculate the N50 score of An Assembly</h2>

<p> N50 score is a measure of how complete a genome assembly is, higher the N50 score, more complete the genome assembly is. It is calculated in a very simple way. First, all the contigs in the assembly are lined up in order of their sequence lengths. Then the contigs are counted, starting from the largest to smallest. When the contig length reaches half of the length of the total assembly, the length of the contig where the counting stopped is the N50 score. Another way of thinking about it is that at least half of all the nucleotides in the assembly are in contigs with N50 length or longer.
Write a Python program **n50.py** that calculates the N50 score of some sequences into an output text file.

<p>The inputs for this program can be generated by the moog.py program. You can run make fasta to create files of 1K, 10K, 100K, and 1M reads in this directory. You can then use these files for testing your program. </p>

The parameters for your program are:

1. One or more positional FILE arguments.
2. An output text file where you output the N50 score (default "n50.txt".

Here is the usage your program should create for -h or --help:

```python
$ ./n50.py -h
usage: n50.py [-h] [-o FILE] FILE [FILE ...]
Calculate N50 score of sequences
positional arguments:
FILE 					 Input FASTA file(s)
optional arguments:
-h, --help               Show this help message and exit
-o FILE, --outfile FILE  Output filename (default: n50.txt)
```

When run with the n1k.fa, it should print this:

```python
$ ./n50.py n1k.fa
File: n1k.fa
Total Size = 
N50 Score = 
```

