<h2> Calculate the N50 score of An Assembly</h2>

Genome assemblies are often given scores to denote how "complete" an assembly is, otherwise, it would be hard to understand the quality of the genome assemble. N50 score is one such measure, higher the N50 score, more complete the genome assembly is. It is calculated in the following way. First, all the contigs in the assembly are lined up in order of their sequence lengths. Then the contig sequences are counted, starting from the largest to smallest until the contig length reaches half of the length of the total assembly. The length of shortest contig in this list is the N50 value. Another way of thinking about it is that at least half of all the nucleotides in the assembly are in contigs of length N50 or higher.
Write a Python program **n50.py** that calculates the N50 score of input sequences into an output text file.

The inputs for this program can be generated by the moog.py program. You can run ***make fasta*** to create fasta files of variable sizes in this directory. You can then use these files for testing your program. There is also another file in the folder- the latest version of the *Arabidopsis thaliana* genome to test the program on an actual genome. The genome is zipped, to save space, so you have to run ***gunzip*** to unzip the genome before running the program on it.

The parameters for your program are:

1. One or more positional FILE arguments.
2. An output text file where you output the N50 score (default "n50.txt").

Here is the usage your program should create for -h or --help:

```python
$ ./n50.py -h
usage: n50.py [-h] [-o FILE] FILE
Calculate N50 score of sequences
positional arguments:
FILE 				Input FASTA file
optional arguments:
-h, --help               Show this help message and exit
-o FILE, --outfile FILE  Output filename (default: n50.txt)
```

When run with **file1.fa** it should print this to the stdout:

```python
$ ./n50.py file1.fa
Processed the fasta file "file1.fa", the N50 Score is 45. Details are in the "n50.txt".
```

The output file should contain the following information:

1. The name of the input file
2. Total size (in base pairs) of the sequences in the fasta file
3. Number of contigs
4. The N50 score.

This is what it should look like :

	Filename: file1.fa
	Total Size = 55,902
	Total Contigs = 456
	N50 Score: 45
All tests should pass:



